src/query/api/v1/handler/placement/common.go

```diff
@@ -36,65 +40,219 @@ import (
+		opts.ServiceZone = v
+	}
+	if v := strings.TrimSpace(headers.Get(HeaderDryRun)); v == "true" {
+		opts.DryRun = true
+	}
+
+	if m3AggOpts != nil {
+		if m3AggOpts.MaxAggregationWindowSize > 0 &&
+			m3AggOpts.MaxAggregationWindowSize > defaultM3AggMaxAggregationWindowSize {
+			opts.M3Agg.MaxAggregationWindowSize = m3AggOpts.MaxAggregationWindowSize
+		}
+
+		if m3AggOpts.WarmupDuration > 0 &&
+			m3AggOpts.MaxAggregationWindowSize > defaultM3AggWarmupDuration {
+			opts.M3Agg.WarmupDuration = m3AggOpts.MaxAggregationWindowSize
+		}
+	}
+
+	return opts
 }
 
 // Service gets a placement service from m3cluster client
-func Service(clusterClient clusterclient.Client, headers http.Header) (placement.Service, error) {
-	ps, _, err := ServiceWithAlgo(clusterClient, headers)
+func Service(
+	clusterClient clusterclient.Client,
+	opts ServiceOptions,
+	now time.Time,
+) (placement.Service, error) {
+	ps, _, err := ServiceWithAlgo(clusterClient, opts, now)
 	return ps, err
 }
 
 // ServiceWithAlgo gets a placement service from m3cluster client and
 // additionally returns an algorithm instance for callers that need fine-grained
 // control over placement updates.
-func ServiceWithAlgo(clusterClient clusterclient.Client, headers http.Header) (placement.Service, placement.Algorithm, error) {
+func ServiceWithAlgo(
+	clusterClient clusterclient.Client,
+	opts ServiceOptions,
+	now time.Time,
+) (placement.Service, placement.Algorithm, error) {
 	cs, err := clusterClient.Services(services.NewOverrideOptions())
 	if err != nil {
 		return nil, nil, err
 	}
 
-	serviceName := DefaultServiceName
-	if v := strings.TrimSpace(headers.Get(HeaderClusterServiceName)); v != "" {
-		serviceName = v
+	if _, ok := allowedServices[opts.ServiceName]; !ok {
+		return nil, nil, fmt.Errorf(
+			"invalid service name: %s, must be one of: %s",
+			opts.ServiceName, allowedServices.String())
 	}
-
-	serviceEnvironment := DefaultServiceEnvironment
-	if v := strings.TrimSpace(headers.Get(HeaderClusterEnvironmentName)); v != "" {
-		serviceEnvironment = v
+	if opts.ServiceName == "" {
+		return nil, nil, errServiceNameIsRequired
 	}
-
-	serviceZone := DefaultServiceZone
-	if v := strings.TrimSpace(headers.Get(HeaderClusterZoneName)); v != "" {
-		serviceZone = v
+	if opts.ServiceEnvironment == "" {
+		return nil, nil, errServiceEnvironmentIsRequired
+	}
+	if opts.ServiceZone == "" {
+		return nil, nil, errServiceZoneIsRequired
+	}
+	if opts.ServiceName == M3AggServiceName && opts.M3Agg == nil {
+		return nil, nil, errM3AggServiceOptionsRequired
 	}
 
 	sid := services.NewServiceID().
-		SetName(serviceName).
-		SetEnvironment(serviceEnvironment).
-		SetZone(serviceZone)
+		SetName(opts.ServiceName).
+		SetEnvironment(opts.ServiceEnvironment).
+		SetZone(opts.ServiceZone)
+
+	pOpts := placement.NewOptions().
+		SetValidZone(opts.ServiceZone).
+		SetIsSharded(true).
+		// M3Agg expects a staged placement.
+		SetIsStaged(true).
```

Comment:
Is this shared by m3db as well? should use `isStaged=false` for m3db and `isStaged=true` for m3agg?

Reply:
not shared, this is inside a if m3agg switch statement

Reply:
oh no its not you're right, will fix